### PR TITLE
Add logic to handle ANSI escape sequences to move cursor

### PR DESCRIPTION
**Pull request by jsbautista**
I made changes to the way the ANSI code \n is handled. Previously, whenever the \n code appeared, a new line was inserted unconditionally. After the change, a new line is inserted unless there is already a written line at the current position. In that case, instead of inserting a new line, the cursor moves one line down.

Preview with changes made:
image

Comparison of the result obtained with the changes made and in Ipython with QTDM

Preview with changes made:

PreviewTQDMQtConsole

Preview with IPython:

PreviewTQDMIpython

Fixes: #350 
Related with: https://github.com/spyder-ide/spyder-kernels/issues/494 https://github.com/spyder-ide/spyder/issues/6172

### CODE REVIEW

**Comment by dalthviz** [comment]
Thanks for your work here @jsbautista ! As quick comment, adding a test for this new move cursor action logic could be worthy

**Comment by TheMatt2** [comment]
> Preview with changes made in #607: 

I appreciate the goal to add these ANSI codes to allow the cursor to move between  lines. This limitation seems to be the main thing standing between qtconsole being a "true terminal emulator", that is compliant'ish with other terminals.

Though, I see some issues with the changes in this MR.

Your GIF seems to show output that is different that what the intended behavior should be. Here is what your `prueba()` prints in VS Code and MacOS's terminal:

 
image

image

"Segunda prueba" should be printed on the second line, as this is printed before any control characters are occur. Not "Segunda prueba" on the last line.


It also looks like this PR breaks `qtconsole/tests/test_00_console_widget.py::TestConsoleWidget::test_print_carriage_return`

**Comment by dalthviz** [comment]
Hi @TheMatt2 thank you for the feedback and giving this work a check! I think the current GIF being shown is a little bit out of date. The main goal of this PR is to give a handling similar to ANSI codes related with cursor movement as you would observe over IPython (but we could indeed give a check to the behavior of other terminals/platforms to contrast possible differences in behavior). So running the `prueba` code with the latest PR commit on Windows you would get the same output as if you run it with IPython:

* Code:

```python
def prueba():
    print("\rPrimera prueba", end="")
    print("\n", end="")
    print("\rSegunda prueba", end="")
    print("\x1b[A", end="")
    print("\n\n", end="")
    print("\rReemplazo", end="")
    print()
```

* This PR:
image

* IPython:
image

Still, we have some edge cases that need handling and also, as you mention, the changes here are causing some tests as the carriage return one to fail but we are giving all that a check.

Hopefully this makes current state of this PR/work clearer but let us know if you have further suggestions/comments/ideas!

**Comment by TheMatt2** [comment]
>  I think the current GIF being shown is a little bit out of date. The main goal of this PR is to give a handling similar to ANSI codes related with cursor movement as you would observe over IPython (but we could indeed give a check to the behavior of other terminals/platforms to contrast possible differences in behavior). ...

Thank you for your response and clarification. Now I'm with you.

File: qtconsole/console_widget.py [hunk 1]
```diff
@@ -2203,7 +2220,13 @@ def _insert_plain_text(self, cursor, text, flush=False):
...
 
                     elif act.action == 'newline':
-                        cursor.movePosition(QtGui.QTextCursor.EndOfLine)
+                        self.log.debug(self._prompt)
+                        if cursor.block() != cursor.document().lastBlock():
+                            cursor.movePosition(QtGui.QTextCursor.NextBlock)
+                        else:
+                            cursor.movePosition(QtGui.QTextCursor.EndOfLine,
+                                                QtGui.QTextCursor.MoveAnchor)
+                            cursor.insertText('\n')
```
Comment (dalthviz) [inline qtconsole/console_widget.py:2229]: @jsbautista I think what is missing here (and the cause of the failed test) is to not only check if we are over the end of the document but also that the document doesn't end with `self._prompt`. So maybe doing something like:

```suggestion
                        if (
                            cursor.block() != cursor.document().lastBlock()
                            and not cursor.document()
                            .toPlainText()
                            .endswith(self._prompt)
                        ):
                            cursor.movePosition(QtGui.QTextCursor.NextBlock)
                        else:
                            cursor.movePosition(
                                QtGui.QTextCursor.EndOfLine,
                                QtGui.QTextCursor.MoveAnchor,
                            )
                            cursor.insertText("\n")
```

Could help tests pass

**Comment by dalthviz** [comment]
Checking aroung some issue related with Rich, using the changes here alone are not enough to see the improvements done. For more details you can check https://github.com/spyder-ide/spyder-kernels/issues/494#issuecomment-2369201113 (where using the Rich Console API is needed - `force_jupyter=False`)

**Review by dalthviz — COMMENTED** [review]
Thanks @jsbautista for your work here! Left a question and a suggestion. Other than that seems like there is not much else to do here, although maybe could be worthy to do some extra manually testing with things like Rich and TQDM (I already did some while checking https://github.com/spyder-ide/spyder-kernels/issues/494)

File: qtconsole/ansi_code_processor.py [hunk 2]
```diff
@@ -185,6 +182,20 @@ def set_csi_code(self, command, params=[]):
             count = params[0] if params else 1
             self.actions.append(ScrollAction('scroll', dir, 'line', count))
 
+        elif (command == 'A'):  # Move N lines Up
```
Comment (ccordoba12) [inline qtconsole/ansi_code_processor.py:185]:
```suggestion
        elif command == 'A':  # Move N lines Up
```

No need for using parenthesis here

File: qtconsole/ansi_code_processor.py [hunk 3]
```diff
@@ -185,6 +182,20 @@ def set_csi_code(self, command, params=[]):
...
+            dir = 'up'
+            count = params[0] if params else 1
+            self.actions.append(MoveAction('move', dir, 'line', count))
+        elif (command == 'B'):  # Move N lines Down
```
Comment (ccordoba12) [inline qtconsole/ansi_code_processor.py:189]:
```suggestion

        elif command == 'B':  # Move N lines Down
```

Follow the format used above these lines.

File: qtconsole/ansi_code_processor.py [hunk 4]
```diff
@@ -185,6 +182,20 @@ def set_csi_code(self, command, params=[]):
...
+            dir = 'down'
+            count = params[0] if params else 1
+            self.actions.append(MoveAction('move', dir, 'line', count))
+        elif (command == 'F'):  # Goes back to the begining of the n-th previous line
```
Comment (ccordoba12) [inline qtconsole/ansi_code_processor.py:193]:
```suggestion

        elif command == 'F':  # Goes back to the begining of the n-th previous line
```

File: qtconsole/console_widget.py [hunk 5]
```diff
@@ -2188,6 +2188,23 @@ def _insert_plain_text(self, cursor, text, flush=False):
                             cursor.select(QtGui.QTextCursor.Document)
                             cursor.removeSelectedText()
 
+                    elif act.action == 'move' and act.unit == 'line':
+                        if act.dir == 'up':
+                            for i in range(act.count):
+                                cursor.movePosition(
+                                    QtGui.QTextCursor.Up)
```
Comment (ccordoba12) [inline qtconsole/console_widget.py:2195]:
```suggestion
                                    QtGui.QTextCursor.Up
                                )
```

Improve code style a bit.

File: qtconsole/console_widget.py [hunk 6]
```diff
@@ -2188,6 +2188,23 @@ def _insert_plain_text(self, cursor, text, flush=False):
...
+                        elif act.dir == 'down':
+                            for i in range(act.count):
+                                cursor.movePosition(
+                                    QtGui.QTextCursor.Down)
```
Comment (ccordoba12) [inline qtconsole/console_widget.py:2199]:
```suggestion
                                    QtGui.QTextCursor.Down
                                )
```

File: qtconsole/console_widget.py [hunk 7]
```diff
@@ -2188,6 +2188,23 @@ def _insert_plain_text(self, cursor, text, flush=False):
...
+                        elif act.dir == 'leftup':
+                            for i in range(act.count):
+                                cursor.movePosition(
+                                    QtGui.QTextCursor.Up)
+                            cursor.movePosition(
+                            QtGui.QTextCursor.StartOfLine,
+                            QtGui.QTextCursor.MoveAnchor)
```
Comment (ccordoba12) [inline qtconsole/console_widget.py:2206]:
```suggestion
                                    QtGui.QTextCursor.Up
                                )
                            cursor.movePosition(
                                QtGui.QTextCursor.StartOfLine,
                                QtGui.QTextCursor.MoveAnchor
                            )
```

File: qtconsole/tests/test_ansi_code_processor.py [hunk 8]
```diff
@@ -182,6 +182,42 @@ def test_combined(self):
         self.assertEqual(splits, ['abc', None, 'def', None])
         self.assertEqual(actions, [[], ['carriage-return'], [], ['backspace']])
 
+    def test_move_cursor_up(self):
+        string = '\x1b[A\x1b[5A\x1b[F\x1b[5F'
```
Comment (ccordoba12) [inline qtconsole/tests/test_ansi_code_processor.py:186]: Please add a comment above this to explain what it does.

**Review by ccordoba12 — COMMENTED** [review]
Thanks @jsbautista for your work on this! Just some small style suggestions and a comment for you.

File: qtconsole/tests/test_ansi_code_processor.py [hunk 9]
```diff
@@ -186,6 +186,10 @@ def test_move_cursor_up(self):
         """Are the ANSI commands for the cursor movement actions
         (movement up and to the beginning of the line) processed correctly?
         """
+        #This line moves the cursor up once, then moves it up five more lines.
+        #Next, it moves the cursor to the beginning of the previous line, and
+        #finally moves it to the beginning of the fifth line above the current 
+        #position
```
Comment (dalthviz) [inline qtconsole/tests/test_ansi_code_processor.py:192]: Lets add a space between the `#` and the text:
```suggestion
        # This line moves the cursor up once, then moves it up five more lines.
        # Next, it moves the cursor to the beginning of the previous line, and
        # finally moves it to the beginning of the fifth line above the current 
        # position
```

**Review by dalthviz — APPROVED** [review]
Thanks @jsbautista for all your work here! This LGTM :+1:

Just in case, do you have any further comments @ccordoba12 ? Also, in case this is ready to be merged, should the millestone be set to be `5.7.0` maybe (since this adds new functionality)?

**Review by ccordoba12 — APPROVED** [review]
Looks good to me now, thanks @jsbautista and @dalthviz for this great improvement!

**Comment by ccordoba12** [comment]
> Also, in case this is ready to be merged, should the millestone be set to be 5.7.0 maybe (since this adds new functionality)?

I think this is closer to be a bug fix than new functionality because it doesn't add new config options, features or shortcuts. As you can see in our [changelog](https://github.com/jupyter/qtconsole/blob/main/docs/source/changelog.rst), we release a new minor version when there are additions like those to the package, or when dropping Python versions.

**Comment by ccordoba12** [comment]
Also, since this basically fixes https://github.com/spyder-ide/spyder/issues/6172, a PR in Spyder should be created to sync our Qtconsole subrepo there with these changes.
